File: api_gateway/src/modules/recommendation/dtos/response/cpu.rs
use serde::Serialize;
use utoipa::ToSchema;

use crate::modules::recommendation::dtos::response::Timestamp;

#[derive(Serialize, ToSchema)]
pub struct CpuDto {
    id: String,
    brand: String,
    r#gen: String,
    family: String,
    series: String,
    cores: i32,
    threads: i32,
    base_clock: f32,
    max_clock: f32,
    cache: i32,
    socket: String,
    graphics: bool,
    oc: bool,
    recommended_power: i32,
    avg_price: f32,
    #[schema(value_type = String, format = DateTime)]
    release_date: Timestamp,
    img: Option<Vec<u8>>,
    #[schema(value_type = String, format = DateTime)]
    created_at: Timestamp,
    #[schema(value_type = Option<String>, format = DateTime)]
    updated_at: Option<Timestamp>,
}

impl CpuDto {
    pub fn new(
        id: String,
        brand: String,
        r#gen: String,
        family: String,
        series: String,
        cores: i32,
        threads: i32,
        base_clock: f32,
        max_clock: f32,
        cache: i32,
        socket: String,
        graphics: bool,
        oc: bool,
        recommended_power: i32,
        avg_price: f32,
        release_date: Timestamp,
        img: Option<Vec<u8>>,
        created_at: Timestamp,
        updated_at: Option<Timestamp>,
    ) -> Self {
        Self {
            id,
            brand,
            r#gen,
            family,
            series,
            cores,
            threads,
            base_clock,
            max_clock,
            cache,
            socket,
            graphics,
            oc,
            recommended_power,
            avg_price,
            release_date,
            img,
            created_at,
            updated_at,
        }
    }
}

File: pc_builder_service/internal/domain/models/cpu.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type CPU struct {
	ID 						uuid.UUID 		`gorm:"type:uuid;primaryKey"`
	Brand 					string 			`gorm:"type:varchar(255);not null"`
	Gen 					string 			`gorm:"type:varchar(255);not null"`
	Family 					string 			`gorm:"type:varchar(255);not null"`
	Series 					string 			`gorm:"type:varchar(255);not null"`
	Cores 					int32 			`gorm:"type:integer;not null"`
	Threads 				int32 			`gorm:"type:integer;not null"`
	BaseClock 				float32 		`gorm:"type:real;not null"`
	MaxClock 				float32 		`gorm:"type:real;not null"`
	Cache 					int32 			`gorm:"type:integer;not null"`
	Socket 					string 			`gorm:"type:varchar(255);not null"`
	Graphics 				bool 			`gorm:"type:boolean;not null"`
	OC 						bool 			`gorm:"type:boolean;not null"`
	RecommendedPower 		int32 			`gorm:"type:integer;not null"`
	ReleaseDate 			time.Time 		`gorm:"type:timestamp;not null"`
	AvgPrice 				float32 		`gorm:"type:real;not null"`
	Img 					[]byte 			`gorm:"type:bytea"`
	CreatedAt 				time.Time 		`gorm:"type:timestamp;not null"`
	UpdatedAt 				time.Time 		`gorm:"type:timestamp;not null"`
}
